<a href="https://colab.research.google.com/github/mbparsons/code/blob/master/Using_LASSO_Regression_to_Model_Fat_Sales_Data.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [0]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

import statsmodels.api as sm

import sklearn as sk
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LassoLarsCV
from statsmodels.stats.outliers_influence import variance_inflation_factor
from sklearn.linear_model import LassoCV, LarsCV, LassoLarsCV
from sklearn.metrics import mean_squared_error, r2_score, mean_absolute_error, accuracy_score

In [0]:
from google.colab import drive
drive.mount('/content/drive')

Go to this URL in a browser: https://accounts.google.com/o/oauth2/auth?client_id=947318989803-6bn6qk8qdgf4n4g3pfee6491hc0brc4i.apps.googleusercontent.com&redirect_uri=urn%3aietf%3awg%3aoauth%3a2.0%3aoob&response_type=code&scope=email%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdocs.test%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive.photos.readonly%20https%3a%2f%2fwww.googleapis.com%2fauth%2fpeopleapi.readonly

Enter your authorization code:
··········
Mounted at /content/drive


In [0]:
ratios = 'drive/My Drive/Colab Notebooks/finalmaster-ratios.csv'
alldata = pd.read_csv(ratios)
alldata

# Purchases  B01001001   B01001002  ...   B19001015  B19001016  B19001017
0             22     206252  469.226965  ...   36.440765  23.446284  21.197485
1              7      61399  486.538869  ...   33.000508  33.169741  24.792689
2              3      73170  489.859232  ...   20.125056  11.890525  16.537397
3             94     251724  505.585483  ...   43.731067  38.851729  40.427349
4              0      37382  495.586111  ...   36.118530  31.603714  19.648989
..           ...        ...         ...  ...         ...        ...        ...
727            4      38106  482.312497  ...   27.706620  19.347445  15.033032
728            3      32531  511.665796  ...  106.653459  84.530296  76.027737
729            3      38761  493.511519  ...   49.390327  47.814769  31.031648
730            1      28359  496.773511  ...   29.939545  24.085980  19.192016
731          507    5861000  484.835011  ...   40.966917  44.925312  52.695060

[732 rows x 190 columns]

In [0]:
allvariablenames = list(alldata.columns.values)
allvariablenames

['# Purchases',
 'B01001001',
 'B01001002',
 'B01001003',
 'B01001004',
 'B01001005',
 'B01001006',
 'B01001007',
 'B01001008',
 'B01001009',
 'B01001010',
 'B01001011',
 'B01001012',
 'B01001013',
 'B01001014',
 'B01001015',
 'B01001016',
 'B01001017',
 'B01001018',
 'B01001019',
 'B01001020',
 'B01001021',
 'B01001022',
 'B01001023',
 'B01001024',
 'B01001025',
 'B01001026',
 'B01001027',
 'B01001028',
 'B01001029',
 'B01001030',
 'B01001031',
 'B01001032',
 'B01001033',
 'B01001034',
 'B01001035',
 'B01001036',
 'B01001037',
 'B01001038',
 'B01001039',
 'B01001040',
 'B01001041',
 'B01001042',
 'B01001043',
 'B01001044',
 'B01001045',
 'B01001046',
 'B01001047',
 'B01001048',
 'B01001049',
 'B02001001',
 'B02001002',
 'B02001003',
 'B02001004',
 'B02001005',
 'B02001006',
 'B02001007',
 'B02001008',
 'B02001009',
 'B02001010',
 'B12001001',
 'B12001002',
 'B12001003',
 'B12001004',
 'B12001005',
 'B12001006',
 'B12001007',
 'B12001008',
 'B12001009',
 'B12001010',
 'B12001011',
 'B1

In [0]:
allvariablenamesnew = allvariablenames[8:]
allvariablenamesnew

['B01001008',
 'B01001009',
 'B01001010',
 'B01001011',
 'B01001012',
 'B01001013',
 'B01001014',
 'B01001015',
 'B01001016',
 'B01001017',
 'B01001018',
 'B01001019',
 'B01001020',
 'B01001021',
 'B01001022',
 'B01001023',
 'B01001024',
 'B01001025',
 'B01001026',
 'B01001027',
 'B01001028',
 'B01001029',
 'B01001030',
 'B01001031',
 'B01001032',
 'B01001033',
 'B01001034',
 'B01001035',
 'B01001036',
 'B01001037',
 'B01001038',
 'B01001039',
 'B01001040',
 'B01001041',
 'B01001042',
 'B01001043',
 'B01001044',
 'B01001045',
 'B01001046',
 'B01001047',
 'B01001048',
 'B01001049',
 'B02001001',
 'B02001002',
 'B02001003',
 'B02001004',
 'B02001005',
 'B02001006',
 'B02001007',
 'B02001008',
 'B02001009',
 'B02001010',
 'B12001001',
 'B12001002',
 'B12001003',
 'B12001004',
 'B12001005',
 'B12001006',
 'B12001007',
 'B12001008',
 'B12001009',
 'B12001010',
 'B12001011',
 'B12001012',
 'B12001013',
 'B12001014',
 'B12001015',
 'B12001016',
 'B12001017',
 'B12001018',
 'B12001019',
 'B130

In [0]:
#load predictors into dataframe
predictors = alldata[allvariablenamesnew]  

#load target into dataframe
target = alldata['# Purchases']   

In [0]:
# split data into train and test sets, with 30% retained for test

pred_train, pred_test, tar_train, tar_test = train_test_split(predictors, target, test_size=.3, random_state=123)   

In [0]:
model = LassoLarsCV(cv=10).fit(predictors, target)

/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/least_angle.py:578: ConvergenceWarning: Regressors in active set degenerate. Dropping a regressor, after 11 iterations, i.e. alpha=4.367e-01, with an active set of 11 regressors, and the smallest cholesky pivot element being 1.054e-08. Reduce max_iter or increase eps parameters.
  ConvergenceWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/least_angle.py:578: ConvergenceWarning: Regressors in active set degenerate. Dropping a regressor, after 44 iterations, i.e. alpha=1.197e-01, with an active set of 40 regressors, and the smallest cholesky pivot element being 1.825e-08. Reduce max_iter or increase eps parameters.
  ConvergenceWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/least_angle.py:578: ConvergenceWarning: Regressors in active set degenerate. Dropping a regressor, after 44 iterations, i.e. alpha=1.187e-01, with an active set of 40 regressors, and the smallest cholesky pivot eleme

In [0]:
#build coefficent chart

predictors_model=pd.DataFrame(allvariablenamesnew)
predictors_model.columns = ['label']
predictors_model['coeff'] = model.coef_

for index, row in predictors_model.iterrows():
    if row['coeff'] > 0:
        print(row.values)

['B01001014' 0.482839698113289]
['B01001036' 1.8818498814890745]
['B01001037' 2.3074573475361975]
['B01001038' 1.9223569348224718]
['B02001005' 0.9541620886147733]
['B13014026' 0.6252570746612429]
['B13014027' 1.429037784828478]
['B13015010' 342242187.1242002]
['B13016008' 0.14173082653108596]
['B15002027' 3.474533859312566]
['B19001013' 0.006867734916745749]
['B19001017' 1.9507395020212406]


In [0]:
predictors_model

label     coeff
0    B01001008  0.000000
1    B01001009  0.000000
2    B01001010  0.000000
3    B01001011  0.000000
4    B01001012  0.000000
..         ...       ...
177  B19001013  0.006868
178  B19001014  0.000000
179  B19001015  0.000000
180  B19001016  0.000000
181  B19001017  1.950740

[182 rows x 2 columns]

Question #1
The lines of code are:
1. turning all the predictor variables into a data frame
2. naming the columns of the data table "label
3. adding a new column with the coefficients from the model 
4. showing all values in rows where the predicted coefficients are not zero, showing which predicted values are significant and useful to predict sales

The variables stand for different age groups and genders from census data.

Question #2

B13015010 and B15002027 most predict sales. In areas where there are more Women Who Did Not Have a Birth in the Past 12 Months and are Now Married (Including Separated and Spouse Absent), we sell more products. Also, where there are more women that are 25 years and over that made it to 12th grade but did not get a diploma, we sell more products.


In [0]:
predictors_model.sort_values(by=['coeff'], ascending=False)

label         coeff
107  B13015010  3.422422e+08
156  B15002027  3.474534e+00
29   B01001037  2.307457e+00
181  B19001017  1.950740e+00
30   B01001038  1.922357e+00
..         ...           ...
51   B02001010 -3.447146e-01
47   B02001006 -8.377936e-01
147  B15002018 -1.973171e+00
108  B13015011 -3.422422e+08
109  B13015012 -3.422422e+08

[182 rows x 2 columns]

Question #3

If I had to report only two census variables to my boss that most steeply predicted sales, I would report B13015010 and B15002027. 

The mean squared error for the training set is 21095.110298440097. The mean squared error for the test set is 36340.928094634444. The training set has a larger mean squared error than the test set.

In [0]:
train_error = mean_squared_error(tar_train, model.predict(pred_train))
print ('training data MSE')
print(train_error)

training data MSE
21095.110298440097


In [0]:
test_error = mean_squared_error(tar_test, model.predict(pred_test))
print ('test data MSE')
print(test_error)

test data MSE
36340.928094634444


Question #4

The training and test set mean squared errors are not similar but are both large numbers. That means practically that both the training set and test set are not very accurate since the MSE's are so large.

R-squared for the training and test set: 

In [0]:
#r squared
rsquared_train=model.score(pred_train,tar_train)
print ('training data R-square')
print(rsquared_train)

training data R-square
0.2721244165386244


In [0]:
#r squared
rsquared_test=model.score(pred_test,tar_test)
print ('test data R-square')
print(rsquared_test)

test data R-square
0.27917606137570605


Question #5

Census data does not predict sales well because the mean squared errors are large and the r-squares are low.

Y-Intercept:

In [0]:
print("y interecept:")
print(model.intercept_)

y interecept:
19.285942627484566


What is our baseline sales number? What does that mean, practically? Think back to what y-intercepts mean in regression models.

Question #6

The baseline sales number is 19.285942627484566. The baseline sales number is the amount of sales without any marketing/before a marketing campaign.